certbot/tests/auth_handler_test.py

```diff
@@ -290,32 +286,46 @@ def test_perform_error(self):
         self.assertEqual(
             self.mock_auth.cleanup.call_args[0][0][0].typ, "tls-sni-01")
 
-    @mock.patch("certbot.auth_handler.AuthHandler._respond")
-    def test_respond_error(self, mock_respond):
+    def test_answer_error(self):
+        self.mock_net.answer_challenge.side_effect = errors.AuthorizationError
+
         authzrs = [gen_dom_authzr(domain="0", challs=acme_util.CHALLENGES)]
         mock_order = mock.MagicMock(authorizations=authzrs)
-        mock_respond.side_effect = errors.AuthorizationError
 
         self.assertRaises(
             errors.AuthorizationError, self.handler.handle_authorizations, mock_order)
         self.assertEqual(self.mock_auth.cleanup.call_count, 1)
         self.assertEqual(
             self.mock_auth.cleanup.call_args[0][0][0].typ, "tls-sni-01")
 
-    @mock.patch("certbot.auth_handler.AuthHandler._poll_challenges")
-    @mock.patch("certbot.auth_handler.AuthHandler.verify_authzr_complete")
-    def test_incomplete_authzr_error(self, mock_verify, mock_poll):
+    def test_incomplete_authzr_error(self):
         authzrs = [gen_dom_authzr(domain="0", challs=acme_util.CHALLENGES)]
         mock_order = mock.MagicMock(authorizations=authzrs)
-        mock_verify.side_effect = errors.AuthorizationError
-        mock_poll.side_effect = self._validate_all
+        self.mock_net.poll.side_effect = _gen_mock_on_poll(status=messages.STATUS_INVALID)
 
-        self.assertRaises(
-            errors.AuthorizationError, self.handler.handle_authorizations, mock_order)
+        with test_util.patch_get_utility():
+            with self.assertRaises(errors.AuthorizationError) as error:
+                self.handler.handle_authorizations(mock_order, False)
+        self.assertTrue('Some challenges have failed' in str(error.exception))
         self.assertEqual(self.mock_auth.cleanup.call_count, 1)
         self.assertEqual(
             self.mock_auth.cleanup.call_args[0][0][0].typ, "tls-sni-01")
 
+    def test_best_effort(self):
```

Comment:
It looks like we didn't have this previously either but since we changed a lot of logic around here, can we add a test where we use `best_effort`, some challenges fail, but we don't raise an exception?

Reply:
I already test that indirectly with this method. On the previous one, I check that the exception raised is "Some challenges have failed", indicating that without best_effort, we have an exception directly in the loop. Here, with best_effort, I test that we raise "All challenges have failed": this is an exception raised outside of the loop.

So this test checks the contract already.

Reply:
It's pretty clear reading the current code and seeing the raised exception that best_effort should work.

I still personally think a unit test where `handle_authorizations` is called with `best_effort=True` and actually returns some but not all of the initial authorizations would be valuable though and may prevent regressions in the future as the code is refactored. I think it's only apparent that `best_effort` works with the given code structure which can change at any time.

I do see your point and we didn't have this test before either so I'll leave it up to you. If you don't think it's worth adding the test, we can skip it.

Reply:
I would lose my DevOps chocolate medal by not responding on the call for a proper unit test. Here is a complete and explicit test for best_effort, with assertions for situations of all authzrs failing and only a part of them failing.